### PR TITLE
Ner

### PR DESCRIPTION
**Pull request by HqWei**

### CODE REVIEW

**Review by yauauau — COMMENTED** [review]
All right so that's how I said all that up

**Comment by innerlee** [comment]
@HqWei linting fails. Remember to install pre-commit :) and sign the cla

**Review by yauauau — COMMENTED** [review]
configs/ner/ner_task.py

**Comment by innerlee** [comment]
These two files are too big. Please keep only a few items. 100-200 lines is acceptable.

image

File: configs/ner/bert_softmax/ner_task.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
+# optimizer
```
Comment (cuhk-hbsun) [inline configs/ner/bert_softmax/ner_task.py]: rename the config file following mm naming style.

File: mmocr/models/ner/classifer/classifier.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
+import torch
+
+from mmdet.models.builder import DETECTORS, build_loss
+from mmocr.models.builder import build_decoder, build_encoder
+from mmocr.models.textrecog.recognizer.base import BaseRecognizer
+
+
+@DETECTORS.register_module()
+class NerClassifier(BaseRecognizer):
+    """Base class for encode-decode recognizer."""
```
Comment (cuhk-hbsun) [inline mmocr/models/ner/classifer/classifier.py:10]: `Base class for NER Classifier`

File: mmocr/models/ner/loss/focal_loss.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import torch
```
Comment (cuhk-hbsun) [inline mmocr/models/ner/loss/focal_loss.py:1]: move this file to `mmocr/models/common/losses`

File: configs/ner/bert_softmax/ner_task_time.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
+# optimizer
```
Comment (cuhk-hbsun) [inline configs/ner/bert_softmax/ner_task_time.py]: rename config file

File: mmocr/core/evaluation/ner.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
+import json
```
Comment (cuhk-hbsun) [inline mmocr/core/evaluation/ner.py:1]: rename to `ner_metric.py`

File: mmocr/core/evaluation/ner.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+from collections import Counter
+
+
+def get_entity_bio(seq, id2label):
+    """Gets entities from sequence.
+    note: BIO
```
Comment (cuhk-hbsun) [inline mmocr/core/evaluation/ner.py:7]: What's the meaning of `BIO`?

File: mmocr/core/evaluation/ner.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
+import json
+from collections import Counter
+
+
+def get_entity_bio(seq, id2label):
```
Comment (cuhk-hbsun) [inline mmocr/core/evaluation/ner.py:5]: add original code source.

File: mmocr/core/evaluation/ner.py [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+                chunks.append(chunk)
+        else:
+            if chunk[2] != -1:
+                chunks.append(chunk)
+            chunk = [-1, -1, -1]
+    return chunks
+
+
+def get_entities(seq, id2label, markup='bios'):
+    '''
```
Comment (cuhk-hbsun) [inline mmocr/core/evaluation/ner.py:47]: reformat docstring

File: mmocr/core/evaluation/ner.py [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+                chunks.append(chunk)
+        else:
+            if chunk[2] != -1:
+                chunks.append(chunk)
+            chunk = [-1, -1, -1]
+    return chunks
+
+
+class SeqEntityScore(object):
+
```
Comment (cuhk-hbsun) [inline mmocr/core/evaluation/ner.py:106]: 1. add code source.
2. docstring

File: mmocr/core/evaluation/ner.py [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+                'f1': round(f1, 4)
+            }
+        origin = len(self.origins)
+        found = len(self.founds)
+        right = len(self.rights)
+        recall, precision, f1 = self.compute(origin, found, right)
+        return {'acc': precision, 'recall': recall, 'f1': f1}, class_info
+
+    def update(self, label_paths, pred_paths):
+        '''
```
Comment (cuhk-hbsun) [inline mmocr/core/evaluation/ner.py:146]: reformat docstring

File: mmocr/core/evaluation/ner.py [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+            line_dict = json.loads(lines[i])
+            text = line_dict['text']
+            label = line_dict['label']
+            label_ids = label2id(label, len(text), label2id_dict, max_len)
+            # break
+            pred = results[i]
+            temp_1 = []
+            temp_2 = []
+
+            for j in range(len(label_ids)):  # enumerate(label_ids):
```
Comment (cuhk-hbsun) [inline mmocr/core/evaluation/ner.py:201]: delete useless code

File: mmocr/datasets/ner_dataset.py [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+        super().__init__(
+            ann_file, loader, pipeline, img_prefix='', test_mode=False)
+        self.ann_file = ann_file
+        self.word2ids = {}
+        self.unknown_id = unknown_id
+        self.start_id = start_id
+        self.end_id = end_id
+        self.max_len = max_len
+        self.map_file = map_file
+        with open(self.map_file) as f:
```
Comment (cuhk-hbsun) [inline mmocr/datasets/ner_dataset.py:49]: `with open(self.map_file, 'r) as f:`

File: mmocr/core/evaluation/ner_metric.py [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+        else:
+            if chunk[2] != -1:
+                chunks.append(chunk)
+            chunk = [-1, -1, -1]
+    return chunks
+
+
+class SeqEntityScore(object):
+    """Get precision, recall and F1-score for named entity recognition task.
+        The code is adapted from https://github.com/lonePatient/BERT-NER-Pytorch
```
Comment (innerlee) [inline mmocr/core/evaluation/ner_metric.py:108]: Add licence info

File: mmocr/core/evaluation/ner_metric.py [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+            self.rights.extend([
+                pre_entity for pre_entity in pre_entities
+                if pre_entity in label_entities
+            ])
+
+
+def label2id(label, text_len, label2id_dict, max_len):
+    id = [0] * max_len
+    for j in range(text_len):
+        id[j] = 31
```
Comment (innerlee) [inline mmocr/core/evaluation/ner_metric.py:171]: Explain the magic number 31

File: mmocr/datasets/ner_dataset.py [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+        """Convert labeled entities to ids.
+
+        Args:
+           label (dict):
+        Returns:
+            list: label ids of an input text
+        """
+        labels = [0] * self.max_len
+        for j in range(text_len + 2):
+            labels[j] = 31
```
Comment (innerlee) [inline mmocr/datasets/ner_dataset.py:88]: Explain the magic number 31

File: tools/train.py [hunk 16]
```diff
@@ -22,6 +22,8 @@ def parse_args():
     parser = argparse.ArgumentParser(description='Train a detector.')
     parser.add_argument('config', help='Train config file path.')
     parser.add_argument('--work-dir', help='The dir to save logs and models.')
+    parser.add_argument(
+        '--load-from', help='The checkpoint file to resume from.')
```
Comment (innerlee) [inline tools/train.py:26]: The help is not exact
